source/common/network/address_impl.cc

```diff
@@ -22,6 +22,35 @@ namespace Address {
 
 namespace {
 
+/**
+ * Convert an IPv4 address to canonical string format.
+ * @note This works similarly to inet_ntop() but is faster.
+ * @param addr address to format.
+ * @return the address in dotted-decimal string format.
+ */
+static std::string Ipv4ToString(const sockaddr_in& addr) {
+  static constexpr size_t BufferSize = 16; // enough space to hold an IPv4 address in string form
+  char str[BufferSize];
+  // Write backwards from the end of the buffer for simplicity.
+  char* start = str + BufferSize;
+  uint32_t ipv4_addr = ntohl(addr.sin_addr.s_addr);
+  for (unsigned i = 4; i != 0; i--, ipv4_addr >>= 8) {
+    uint32_t octet = ipv4_addr & 0xff;
+    if (octet == 0) {
+      *--start = '0';
+    } else {
+      do {
+        *--start = '0' + (octet % 10);
+        octet /= 10;
+      } while (octet != 0);
```

Comment:
Also worth trying for speed: a 256 element static lookup table. I would declare this as:
```
  static const char* OctedLookup[] = {"0", "1", "2", ...};
```
to get the compiler to lay it out in text-space, as opposed to having it constructed at runtime (and also running into static-initialization fiasco issues).

Reply:
did you get a chance to try this?

Reply:
I won't have time to try a lookup-table implementation, but I won't object if anybody else wants to make a follow-up PR for it.